### DISCUSSION TITLE
Streamlines as "ragged" arrays

### DISCUSSION THREAD

**Discussion by arokem** [dipy/dipy#3244] [3. Community Lounge]
Our streamlines data is a ["jagged" or "ragged" array](https://en.wikipedia.org/wiki/Jagged_array). Of potential interest here, scientific programmers in other domains have done extensive work on "ragged arrays" including this interesting library out of the high-energy physics world: https://github.com/scikit-hep/ragged. One of the interesting and potentially useful aspects of this is that it follows [the array API specification](https://data-apis.org/array-api/latest/API_specification), which means that this will (eventually) be interoperable with multiple different array libraries (e.g., jax, pytorch, etc.). 

I am wondering if we should experiment with this approach to see whether it saves time/memory for some of the operations that we do with streamlines.

**Comment by skoudoro** [comment]
the [ArraySequence](https://github.com/nipy/nibabel/blob/e6ccec4dcda98911f6ad7613ac2069736e25b144/nibabel/streamlines/array_sequence.py#L102) object has been created to avoid ragged array and memory issue if I am not wrong. So the ArraySequence is never a ragged array since we have:
- a list of points Nx3 containing all streamlines
- an array of offset defining the beginning + end of the streamlines

We tried to enforce the use of ArraySequence (that we call Streamlines) and I am pretty sure that this object can be easily compatible with [the array API specification](https://data-apis.org/array-api/latest/API_specification).

our `StatefulTractogram` is using the `ArraySequence` object.  

We should remove in our codebase any Streamline defined as list of `np.array`, or nested `np.array` to avoid ragged array and memory issue.

**Reply by arokem** [reply]
I realize that under the hood our representation of streamlines is not a ragged array, but the data structure is conceptually still a ragged array. So, we could potentially benefit from work that others have done to make computing with these kinds of data structures more efficient and straightforward. 

[Here](https://awkward-array.org/doc/main/index.html)'s another library that seems to provide both ways to do [numpy operations](https://awkward-array.org/doc/main/user-guide/how-to-math-numpy.html), but also some [interesting interactions with dask](https://dask-awkward.readthedocs.io/en/stable/). 

It might be interesting to experiment with that at some point.
